### PR TITLE
feat: support non-struct type class structure

### PR DESCRIPTION
**Pull request by jvanstraten**
This implements and documents the things discussed in https://substrait.slack.com/archives/C02D7CTQXHD/p1662542471566549.

### CODE REVIEW

**Review by jacques-n — APPROVED** [review]
Lgtm. Thanks for formalizing this @jvanstraten

File: site/docs/types/type_classes.md [hunk 1]
```diff
@@ -44,18 +44,42 @@ Compound type classes are type classes that need to be configured by means of a
...
+
+The name-type object notation used above is syntactic sugar for `NSTRUCT<longitude: i32, latitude: i32>`. The following means the same thing:
+
+```yaml
+name: point
+structure: "NSTRUCT<longitude: i32, latitude: i32>"
 ```
 
-This declares a new type (namespaced to the associated YAML file) called "point". This type is composed of two `i32` values named longitude and latitude. Once a type has been declared, it can be used in function declarations.  [TBD: should field references be allowed to dereference the components of a user defined type?]
+The structure field of a type is only intended to inform systems that don't have built-in support for the type how they can transfer the data type from one point to another without unnecessarily serialization/deserialization *and* without loss of type safety. Note that it is currently not possible to "unpack" a user-defined type class into its structure type or components thereof without also defining a function to do that packing/unpacking.
```
Comment (westonpace) [inline site/docs/types/type_classes.md:69]:
```suggestion
The structure field of a type is only intended to inform systems that don't have built-in support for the type how they can transfer the data type from one point to another without unnecessary serialization/deserialization *and* without loss of type safety. Note that it is currently not possible to "unpack" a user-defined type class into its structure type or components thereof without also defining a function to do that packing/unpacking.
```

Small typo but also, I don't really understand this last sentence.  Can you expand on what you mean here?

Comment (jvanstraten) [reply site/docs/types/type_classes.md]: Fixed typo and reworded sentence in ad75c2a. Does it make more sense when written that way? The point is to get rid of the TBD w.r.t. whether FieldRefs can index into non-opaque user-defined types.

Comment (westonpace) [reply site/docs/types/type_classes.md]: Yes, thank you.  The new wording is clear.

**Review by westonpace — APPROVED** [review]

**Pull request merged**